File: flutter_app/lib/utils/loan_utils.dart
import 'dart:math';
import 'package:intl/intl.dart';
import '../models/borrower.dart';
import '../models/credit_assessment.dart';
import '../models/loan.dart';
import '../models/payment.dart';
import '../models/schedule_installment.dart';

class LoanStats {
  final double totalDisbursed;
  final double totalScheduled;
  final double totalPaid;
  final double outstandingBalance;
  final double overdueAmount;
  final double penaltyAmount;
  final double totalDueWithPenalty;
  final double creditBalance;
  final double payoffAmount;
  final int progressPct;
  final ScheduleInstallment? nextDue;
  final List<ScheduleInstallment> scheduleWithStatus;

  LoanStats({
    required this.totalDisbursed,
    required this.totalScheduled,
    required this.totalPaid,
    required this.outstandingBalance,
    required this.overdueAmount,
    required this.penaltyAmount,
    required this.totalDueWithPenalty,
    this.creditBalance = 0.0,
    required this.payoffAmount,
    required this.progressPct,
    this.nextDue,
    required this.scheduleWithStatus,
  });
}

class LoanUtils {
  static String defaultCurrencyCode = 'PHP';
  static String defaultDateFormat = 'MMM d, yyyy';

  static double round2(double val) {
    return (val * 100.0).round() / 100.0;
  }

  static double calculateUpfrontDeduction(double principal, String type, double value) {
    final p = max(0.0, principal);
    final val = max(0.0, value);

    if (type == 'percent') {
      return round2(p * (val / 100.0));
    } else if (type == 'fixed') {
      return round2(min(p, val));
    }
    return 0.0;
  }

  static double calculateNetDisbursed(double principal, String type, double value) {
    final deduction = calculateUpfrontDeduction(principal, type, value);
    return round2(max(0.0, principal - deduction));
  }

  static String currencySymbol([String? currencyCode]) {
    final code = currencyCode ?? defaultCurrencyCode;
    switch (code.toUpperCase()) {
      case 'EUR':
        return '€';
      case 'PHP':
        return '₱';
      case 'GBP':
        return '£';
      case 'USD':
      default:
        return '\$';
    }
  }

  static String formatCurrency(double amount, [String? currencyCode]) {
    final code = currencyCode ?? defaultCurrencyCode;
    final symbol = currencySymbol(code);
    final formatter = NumberFormat.currency(locale: 'en_US', symbol: symbol, decimalDigits: 2);
    return formatter.format(amount);
  }

  static String formatDate(String? dateStr, [String? formatPattern]) {
    if (dateStr == null || dateStr.isEmpty) return 'N/A';
    final pattern = formatPattern ?? defaultDateFormat;
    try {
      final dt = DateTime.parse(dateStr);
      return DateFormat(pattern).format(dt);
    } catch (_) {
      return dateStr;
    }
  }

  static String formatPercent(double rate) {
    return '${rate.toStringAsFixed(1)}%';
  }

  static DateTime calculateDueDate(DateTime startDate, String frequency, int periodIndex) {
    switch (frequency) {
      case 'daily':
        return startDate.add(Duration(days: periodIndex));
      case 'weekly':
        return startDate.add(Duration(days: periodIndex * 7));
      case 'biweekly':
        return startDate.add(Duration(days: periodIndex * 14));
      case 'monthly':
      default:
        var year = startDate.year;
        var month = startDate.month + periodIndex;
        while (month > 12) {
          month -= 12;
          year += 1;
        }
        final day = min(startDate.day, 28);
        return DateTime(year, month, day);
    }
  }

  static List<ScheduleInstallment> generateSchedule(
    double principal,
    double interestRate,
    int termCount,
    String disbursementDate, {
    String repaymentFrequency = 'monthly',
    String interestMethod = 'reducing',
  }) {
    final p = max(0.0, principal);
    final rate = max(0.0, interestRate);
    final isOneTime = interestMethod == 'one_time';
    final n = isOneTime ? 1 : max(1, termCount);

    DateTime startDate;
    try {
      startDate = disbursementDate.isNotEmpty ? DateTime.parse(disbursementDate) : DateTime.now();
    } catch (_) {
      startDate = DateTime.now();
    }

    final List<ScheduleInstallment> schedule = [];

    if (interestMethod == 'flat') {
      // Note: Flat/Add-on ("5-6") interest is intentionally period-independent (total interest = principal * rate / 100) regardless of term length.
      final totalInterest = p * (rate / 100.0);
      final principalPerPeriod = p / n;
      final interestPerPeriod = totalInterest / n;
      double balance = p;

      for (int i = 1; i <= n; i++) {
        final dueDate = calculateDueDate(startDate, repaymentFrequency, i);
        final dueDateStr = DateFormat('yyyy-MM-dd').format(dueDate);

        double prin = principalPerPeriod;
        double instInterest = interestPerPeriod;

        if (i == n) {
          prin = balance;
          balance = 0.0;
        } else {
          balance -= prin;
        }

        schedule.add(ScheduleInstallment(
          installmentNo: i,
          dueDate: dueDateStr,
          amount: round2(prin + instInterest),
          principal: round2(prin),
          interest: round2(instInterest),
          balance: max(0.0, round2(balance)),
        ));
      }
      return schedule;
    }

    if (interestMethod == 'interest_only') {
      double perPeriodRate = 0.0;
      switch (repaymentFrequency) {
        case 'daily':
          perPeriodRate = (rate / 100.0) / 365.0;
          break;
        case 'weekly':
          perPeriodRate = (rate / 100.0) / 52.0;
          break;
        case 'biweekly':
          perPeriodRate = (rate / 100.0) / 26.0;
          break;
        case 'monthly':
        default:
          perPeriodRate = (rate / 100.0) / 12.0;
          break;
      }

      final interestPerPeriod = p * perPeriodRate;
      double balance = p;

      for (int i = 1; i <= n; i++) {
        final dueDate = calculateDueDate(startDate, repaymentFrequency, i);
        final dueDateStr = DateFormat('yyyy-MM-dd').format(dueDate);

        final prin = (i == n) ? p : 0.0;
        if (i == n) balance = 0.0;

        schedule.add(ScheduleInstallment(
          installmentNo: i,
          dueDate: dueDateStr,
          amount: round2(prin + interestPerPeriod),
          principal: round2(prin),
          interest: round2(interestPerPeriod),
          balance: max(0.0, round2(balance)),
        ));
      }
      return schedule;
    }

    if (interestMethod == 'one_time') {
      // Note: One-time payment interest is intentionally period-independent.
      final dueDate = calculateDueDate(startDate, repaymentFrequency, 1);
      final dueDateStr = DateFormat('yyyy-MM-dd').format(dueDate);
      final totalInterest = p * (rate / 100.0);

      schedule.add(ScheduleInstallment(
        installmentNo: 1,
        dueDate: dueDateStr,
        amount: round2(p + totalInterest),
        principal: round2(p),
        interest: round2(totalInterest),
        balance: 0.0,
      ));
      return schedule;
    }

    // Default: 'reducing' (amortizing balance)
    double perPeriodRate = 0.0;
    switch (repaymentFrequency) {
      case 'daily':
        perPeriodRate = (rate / 100.0) / 365.0;
        break;
      case 'weekly':
        perPeriodRate = (rate / 100.0) / 52.0;
        break;
      case 'biweekly':
        perPeriodRate = (rate / 100.0) / 26.0;
        break;
      case 'monthly':
      default:
        perPeriodRate = (rate / 100.0) / 12.0;
        break;
    }

    final r = perPeriodRate;
    double periodPayment = 0.0;
    if (r > 0) {
      periodPayment = (p * r * pow(1 + r, n)) / (pow(1 + r, n) - 1);
    } else {
      periodPayment = p / n;
    }

    double balance = p;

    for (int i = 1; i <= n; i++) {
      final dueDate = calculateDueDate(startDate, repaymentFrequency, i);
      final dueDateStr = DateFormat('yyyy-MM-dd').format(dueDate);

      final interestForPeriod = balance * r;
      double principalForPeriod = periodPayment - interestForPeriod;

      if (i == n || (balance - principalForPeriod) < 0.01) {
        principalForPeriod = balance;
        balance = 0.0;
      } else {
        balance -= principalForPeriod;
      }

      final installmentAmount = principalForPeriod + interestForPeriod;

      schedule.add(ScheduleInstallment(
        installmentNo: i,
        dueDate: dueDateStr,
        amount: round2(installmentAmount),
        principal: round2(principalForPeriod),
        interest: round2(interestForPeriod),
        balance: max(0.0, round2(balance)),
      ));
    }

    return schedule;
  }

  static List<ScheduleInstallment> getScheduleWithStatus(
    List<ScheduleInstallment> schedule,
    List<Payment> payments, [
    String loanStatus = 'active',
    DateTime? referenceDate,
  ]) {
    final refDate = referenceDate ?? DateTime.now();
    final cutoffDate = DateTime(refDate.year, refDate.month, refDate.day, 23, 59, 59, 999);

    double availablePayment = payments.fold(0.0, (sum, p) => sum + p.amount);

    return schedule.map((inst) {
      final instAmount = inst.amount;
      double paidAmount = 0.0;
      String status = 'pending';

      if (availablePayment >= instAmount) {
        paidAmount = instAmount;
        availablePayment -= instAmount;
        status = 'paid';
      } else if (availablePayment > 0) {
        paidAmount = availablePayment;
        availablePayment = 0.0;
        status = 'partial';
      } else {
        paidAmount = 0.0;
      }

      if (status != 'paid') {
        try {
          final due = DateTime.parse(inst.dueDate);
          final dueCutoff = DateTime(due.year, due.month, due.day, 23, 59, 59, 999);
          if (dueCutoff.isBefore(cutoffDate)) {
            status = 'overdue';
          }
        } catch (_) {}
      }

      if (loanStatus == 'rejected') {
        status = 'cancelled';
      } else if (loanStatus == 'defaulted' && status != 'paid') {
        status = 'overdue';
      }

      final remainingAmount = max(0.0, round2(instAmount - paidAmount));

      return inst.copyWith(
        paidAmount: round2(paidAmount),
        remainingAmount: remainingAmount,
        status: status,
      );
    }).toList();
  }

  static double calculatePenalty(Loan loan, List<ScheduleInstallment> scheduleWithStatus, DateTime referenceDate) {
    final type = loan.penaltyType;
    final val = max(0.0, loan.penaltyValue);
    final accrued = max(0.0, loan.accruedPenalty);

    if (type == 'none' || val == 0.0) return accrued;

    final overdueInsts = scheduleWithStatus.where((inst) => inst.status == 'overdue').toList();

    double newlyIncurred = 0.0;
    if (overdueInsts.isNotEmpty) {
      if (type == 'percent_per_period') {
        for (final inst in overdueInsts) {
          newlyIncurred += inst.remainingAmount * (val / 100.0);
        }
      } else if (type == 'fixed_per_period') {
        newlyIncurred = val * overdueInsts.length;
      } else if (type == 'fixed_once') {
        // fixed_once applies at most once over the life of the loan
        newlyIncurred = accrued > 0 ? 0.0 : val;
      }
    }

    return round2(accrued + newlyIncurred);
  }

  static double computeEarlyPayoffAmount(Loan loan, [DateTime? asOfDate]) {
    final stats = getLoanStats(loan, asOfDate);
    if (loan.interestMethod != 'reducing') {
      return stats.totalDueWithPenalty;
    }

    double remainingPrincipal = 0.0;
    for (final inst in stats.scheduleWithStatus) {
      if (inst.status != 'paid') {
        final paidPct = inst.amount > 0 ? (inst.paidAmount / inst.amount) : 0.0;
        final unpaidPrin = inst.principal * (1.0 - paidPct);
        remainingPrincipal += unpaidPrin;
      }
    }
    remainingPrincipal = round2(remainingPrincipal);

    if (remainingPrincipal <= 0) {
      return round2(stats.penaltyAmount);
    }

    DateTime lastDate;
    try {
      if (loan.payments.isNotEmpty) {
        lastDate = DateTime.parse(loan.payments.last.date);
      } else if (loan.disbursementDate.isNotEmpty) {
        lastDate = DateTime.parse(loan.disbursementDate);
      } else {
        lastDate = DateTime.now();
      }
    } catch (_) {
      lastDate = DateTime.now();
    }

    final refDate = asOfDate ?? DateTime.now();
    final daysElapsed = max(0, refDate.difference(lastDate).inDays);

    final dailyRate = (loan.interestRate / 100.0) / 365.0;
    final accruedInterest = round2(remainingPrincipal * dailyRate * daysElapsed);

    final payoff = round2(remainingPrincipal + accruedInterest + stats.penaltyAmount);
    return min(payoff, stats.totalDueWithPenalty);
  }

  static String? validateLoanParams({
    required double principal,
    required double interestRate,
    required int termCount,
    required String repaymentFrequency,
    required double penaltyValue,
  }) {
    if (principal <= 0) return 'Principal must be greater than 0.';
    if (principal > 10000000) return 'Principal exceeds maximum allowed limit (₱10,000,000).';
    if (interestRate < 0 || interestRate > 100) return 'Interest rate must be between 0% and 100%.';
    if (penaltyValue < 0) return 'Penalty value cannot be negative.';

    int maxTerms = 60;
    switch (repaymentFrequency) {
      case 'daily':
        maxTerms = 365;
        break;
      case 'weekly':
        maxTerms = 104;
        break;
      case 'biweekly':
        maxTerms = 52;
        break;
      case 'monthly':
      default:
        maxTerms = 60;
        break;
    }

    if (termCount <= 0 || termCount > maxTerms) {
      return 'Term count for $repaymentFrequency frequency must be between 1 and $maxTerms.';
    }

    return null;
  }

  static LoanStats getLoanStats(Loan loan, [DateTime? referenceDate]) {
    final payments = loan.payments;
    final schedule = loan.schedule;

    final totalPaid = payments.fold(0.0, (sum, p) => sum + p.amount);
    final totalScheduled = schedule.fold(0.0, (sum, s) => sum + s.amount) > 0
        ? schedule.fold(0.0, (sum, s) => sum + s.amount)
        : loan.principal;

    final scheduleWithStatus = getScheduleWithStatus(schedule, payments, loan.status, referenceDate);

    final outstandingBalance = scheduleWithStatus.fold(0.0, (sum, inst) => sum + inst.remainingAmount);

    final overdueAmount = scheduleWithStatus
        .where((inst) => inst.status == 'overdue')
        .fold(0.0, (sum, inst) => sum + inst.remainingAmount);

    final progressPct = totalScheduled > 0
        ? min(100, ((totalPaid / totalScheduled) * 100).round())
        : 0;

    ScheduleInstallment? nextDue;
    try {
      nextDue = scheduleWithStatus.firstWhere(
        (inst) => inst.status == 'pending' || inst.status == 'overdue' || inst.status == 'partial',
      );
    } catch (_) {
      nextDue = null;
    }

    final netDisbursed = calculateNetDisbursed(
      loan.principal,
      loan.upfrontDeductionType,
      loan.upfrontDeductionValue,
    );

    final refDate = referenceDate ?? DateTime.now();
    final penaltyAmount = calculatePenalty(loan, scheduleWithStatus, refDate);
    final totalDueWithPenalty = round2(outstandingBalance + penaltyAmount);

    final totalRequired = round2(totalScheduled + penaltyAmount);
    final creditBalance = totalPaid > totalRequired ? round2(totalPaid - totalRequired) : 0.0;

    // Compute early payoff figure
    double payoff = totalDueWithPenalty;
    if (loan.interestMethod == 'reducing') {
      double remainingPrincipal = 0.0;
      for (final inst in scheduleWithStatus) {
        if (inst.status != 'paid') {
          final paidPct = inst.amount > 0 ? (inst.paidAmount / inst.amount) : 0.0;
          final unpaidPrin = inst.principal * (1.0 - paidPct);
          remainingPrincipal += unpaidPrin;
        }
      }
      remainingPrincipal = round2(remainingPrincipal);

      if (remainingPrincipal <= 0) {
        payoff = round2(penaltyAmount);
      } else {
        DateTime lastDate;
        try {
          if (loan.payments.isNotEmpty) {
            lastDate = DateTime.parse(loan.payments.last.date);
          } else if (loan.disbursementDate.isNotEmpty) {
            lastDate = DateTime.parse(loan.disbursementDate);
          } else {
            lastDate = DateTime.now();
          }
        } catch (_) {
          lastDate = DateTime.now();
        }

        final daysElapsed = max(0, refDate.difference(lastDate).inDays);
        final dailyRate = (loan.interestRate / 100.0) / 365.0;
        final accruedInterest = round2(remainingPrincipal * dailyRate * daysElapsed);

        payoff = min(round2(remainingPrincipal + accruedInterest + penaltyAmount), totalDueWithPenalty);
      }
    }

    return LoanStats(
      totalDisbursed: netDisbursed,
      totalScheduled: round2(totalScheduled),
      totalPaid: round2(totalPaid),
      outstandingBalance: round2(outstandingBalance),
      overdueAmount: round2(overdueAmount),
      penaltyAmount: penaltyAmount,
      totalDueWithPenalty: totalDueWithPenalty,
      creditBalance: creditBalance,
      payoffAmount: payoff,
      progressPct: progressPct,
      nextDue: nextDue,
      scheduleWithStatus: scheduleWithStatus,
    );
  }

  static CreditAssessment assessBorrower(Borrower borrower, List<Loan> borrowerLoans) {
    final baseCreditScore = borrower.creditScore;
    final monthlyIncome = borrower.monthlyIncome;

    final activeLoans = borrowerLoans.where((l) => l.status == 'active').toList();
    double monthlyDebt = 0.0;

    for (final loan in activeLoans) {
      double periodAmount = 0.0;
      if (loan.schedule.isNotEmpty) {
        periodAmount = loan.schedule[0].amount;
      } else if (loan.principal > 0 && loan.termCount > 0) {
        periodAmount = loan.principal / loan.termCount;
      }

      double multiplier = 1.0;
      switch (loan.repaymentFrequency) {
        case 'daily':
          multiplier = 30.4167; // 365 / 12
          break;
        case 'weekly':
          multiplier = 4.3333; // 52 / 12
          break;
        case 'biweekly':
          multiplier = 2.1667; // 26 / 12
          break;
        case 'monthly':
        default:
          multiplier = 1.0;
          break;
      }

      monthlyDebt += periodAmount * multiplier;
    }

    final dtiPct = monthlyIncome > 0
        ? ((monthlyDebt / monthlyIncome) * 100).round()
        : (monthlyDebt > 0 ? 100 : 0);

    final completedLoans = borrowerLoans.where((l) => l.status == 'completed').toList();
    final defaultedLoans = borrowerLoans.where((l) => l.status == 'defaulted').toList();

    int scoreAdjustment = 0;
    scoreAdjustment += min(30, completedLoans.length * 10);
    scoreAdjustment -= defaultedLoans.length * 25;

    if (dtiPct > 50) {
      scoreAdjustment -= 15;
    } else if (dtiPct > 35) {
      scoreAdjustment -= 5;
    }

    final derivedScore = max(0, min(100, baseCreditScore + scoreAdjustment));

    String riskRating = 'medium';
    if (derivedScore >= 70 && dtiPct <= 35) {
      riskRating = 'low';
    } else if (derivedScore < 45 || dtiPct > 50 || defaultedLoans.isNotEmpty) {
      riskRating = 'high';
    }

    return CreditAssessment(
      creditScore: derivedScore,
      baseCreditScore: baseCreditScore,
      dtiPct: dtiPct,
      riskRating: riskRating,
      monthlyDebt: round2(monthlyDebt),
      completedCount: completedLoans.length,
      defaultedCount: defaultedLoans.length,
      activeCount: activeLoans.length,
    );
  }
}

File: src/utils/loanUtils.js
/**
 * Utility functions for MicroLend loan calculations, schedule generation,
 * borrower risk assessment, and formatting.
 */

/**
 * Formats a number as USD currency.
 */
export function formatCurrency(amount) {
  const num = Number(amount) || 0;
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(num);
}

/**
 * Formats a date string into MMM DD, YYYY format.
 */
export function formatDate(dateStr) {
  if (!dateStr) return 'N/A';
  const date = new Date(dateStr);
  if (isNaN(date.getTime())) return dateStr;
  return new Intl.DateTimeFormat('en-US', {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
  }).format(date);
}

/**
 * Formats a number as a percentage string.
 */
export function formatPercent(rate) {
  const num = Number(rate) || 0;
  return `${num.toFixed(1)}%`;
}

/**
 * Generates an amortizing loan schedule using standard equal-monthly-payment formula.
 * @param {number} principal - Loan amount
 * @param {number} interestRateAnnual - Annual interest rate percentage (e.g. 12 for 12%)
 * @param {number} termMonths - Loan duration in months
 * @param {string} disbursementDate - Start / disbursement date string (YYYY-MM-DD)
 * @returns {Array} Array of schedule installments
 */
export function generateSchedule(principal, interestRateAnnual, termMonths, disbursementDate) {
  const p = Math.max(0, Number(principal) || 0);
  const rate = Math.max(0, Number(interestRateAnnual) || 0);
  const n = Math.max(1, parseInt(termMonths, 10) || 1);
  const startDate = disbursementDate ? new Date(disbursementDate) : new Date();

  const r = (rate / 100) / 12;

  let monthlyPayment = 0;
  if (r > 0) {
    monthlyPayment = (p * r * Math.pow(1 + r, n)) / (Math.pow(1 + r, n) - 1);
  } else {
    monthlyPayment = p / n;
  }

  let balance = p;
  const schedule = [];

  for (let i = 1; i <= n; i++) {
    const dueDate = new Date(startDate);
    dueDate.setMonth(dueDate.getMonth() + i);
    const dateStr = dueDate.toISOString().split('T')[0];

    const interestForMonth = balance * r;
    let principalForMonth = monthlyPayment - interestForMonth;

    // Handle last installment edge cases / rounding drift
    if (i === n || balance - principalForMonth < 0.01) {
      principalForMonth = balance;
      balance = 0;
    } else {
      balance = balance - principalForMonth;
    }

    const installmentAmount = principalForMonth + interestForMonth;

    schedule.push({
      installmentNo: i,
      due_date: dateStr,
      amount: Math.round(installmentAmount * 100) / 100,
      principal: Math.round(principalForMonth * 100) / 100,
      interest: Math.round(interestForMonth * 100) / 100,
      balance: Math.max(0, Math.round(balance * 100) / 100),
    });
  }

  return schedule;
}

/**
 * Evaluates payment history against schedule to assign status to each installment.
 * Statuses: 'paid', 'partial', 'overdue', 'pending'
 */
export function getScheduleWithStatus(schedule = [], payments = [], loanStatus = 'active', referenceDate = new Date()) {
  const refDate = new Date(referenceDate);
  refDate.setHours(23, 59, 59, 999);

  // Total paid available to allocate sequentially
  let availablePayment = (payments || []).reduce((sum, p) => sum + (Number(p.amount) || 0), 0);

  return (schedule || []).map((inst) => {
    const instAmount = Number(inst.amount) || 0;
    let paidAmount = 0;
    let status = 'pending';

    if (availablePayment >= instAmount) {
      paidAmount = instAmount;
      availablePayment -= instAmount;
      status = 'paid';
    } else if (availablePayment > 0) {
      paidAmount = availablePayment;
      availablePayment = 0;
      status = 'partial';
    } else {
      paidAmount = 0;
    }

    if (status !== 'paid') {
      const dueDate = new Date(inst.due_date);
      dueDate.setHours(23, 59, 59, 999);

      if (dueDate < refDate) {
        status = 'overdue';
      }
    }

    // Override if loan was rejected or completed or defaulted
    if (loanStatus === 'rejected') {
      status = 'cancelled';
    } else if (loanStatus === 'defaulted' && status !== 'paid') {
      status = 'overdue'; // or defaulted
    }

    const remainingAmount = Math.max(0, Math.round((instAmount - paidAmount) * 100) / 100);

    return {
      ...inst,
      paidAmount: Math.round(paidAmount * 100) / 100,
      remainingAmount,
      status,
    };
  });
}

/**
 * Computes aggregated statistics for a single loan.
 */
export function getLoanStats(loan) {
  if (!loan) return {};

  const payments = loan.payments || [];
  const schedule = loan.schedule || [];

  const totalPaid = payments.reduce((sum, p) => sum + (Number(p.amount) || 0), 0);
  const totalScheduled = schedule.reduce((sum, s) => sum + (Number(s.amount) || 0), 0) || Number(loan.principal) || 0;

  const scheduleWithStatus = getScheduleWithStatus(schedule, payments, loan.status);

  const outstandingBalance = scheduleWithStatus.reduce((sum, inst) => {
    return sum + (inst.remainingAmount || 0);
  }, 0);

  const overdueAmount = scheduleWithStatus
    .filter((inst) => inst.status === 'overdue')
    .reduce((sum, inst) => sum + inst.remainingAmount, 0);

  const progressPct = totalScheduled > 0 ? Math.min(100, Math.round((totalPaid / totalScheduled) * 100)) : 0;

  const nextDue = scheduleWithStatus.find((inst) => inst.status === 'pending' || inst.status === 'overdue' || inst.status === 'partial');

  return {
    totalDisbursed: Number(loan.principal) || 0,
    totalScheduled: Math.round(totalScheduled * 100) / 100,
    totalPaid: Math.round(totalPaid * 100) / 100,
    outstandingBalance: Math.round(outstandingBalance * 100) / 100,
    overdueAmount: Math.round(overdueAmount * 100) / 100,
    progressPct,
    nextDue: nextDue ? { due_date: nextDue.due_date, amount: nextDue.remainingAmount, status: nextDue.status } : null,
    scheduleWithStatus,
  };
}

/**
 * Assesses a borrower's risk rating based on income, active loan obligations,
 * credit score, and payment track record.
 */
export function assessBorrower(borrower, borrowerLoans = []) {
  if (!borrower) return { creditScore: 50, dtiPct: 0, riskRating: 'medium' };

  const baseCreditScore = Number(borrower.credit_score) || 50;
  const monthlyIncome = Number(borrower.monthly_income) || 0;

  // Active loans monthly payment obligation
  const activeLoans = borrowerLoans.filter((l) => l.status === 'active');
  let monthlyDebt = 0;

  activeLoans.forEach((loan) => {
    if (loan.schedule && loan.schedule.length > 0) {
      monthlyDebt += Number(loan.schedule[0].amount) || 0;
    } else if (loan.principal && loan.term_months) {
      monthlyDebt += Number(loan.principal) / Number(loan.term_months);
    }
  });

  const dtiPct = monthlyIncome > 0 ? Math.round((monthlyDebt / monthlyIncome) * 100) : (monthlyDebt > 0 ? 100 : 0);

  const completedLoans = borrowerLoans.filter((l) => l.status === 'completed');
  const defaultedLoans = borrowerLoans.filter((l) => l.status === 'defaulted');

  // Adjust score based on history
  let scoreAdjustment = 0;
  scoreAdjustment += Math.min(30, completedLoans.length * 10);
  scoreAdjustment -= defaultedLoans.length * 25;

  if (dtiPct > 50) scoreAdjustment -= 15;
  else if (dtiPct > 35) scoreAdjustment -= 5;

  const derivedScore = Math.max(0, Math.min(100, baseCreditScore + scoreAdjustment));

  let riskRating = 'medium';
  if (derivedScore >= 70 && dtiPct <= 35) {
    riskRating = 'low';
  } else if (derivedScore < 45 || dtiPct > 50 || defaultedLoans.length > 0) {
    riskRating = 'high';
  }

  return {
    creditScore: derivedScore,
    baseCreditScore,
    dtiPct,
    riskRating,
    monthlyDebt: Math.round(monthlyDebt * 100) / 100,
    completedCount: completedLoans.length,
    defaultedCount: defaultedLoans.length,
    activeCount: activeLoans.length,
  };
}
